google/cloud/bigtable/internal/async_retry_unary_rpc.h

```diff
@@ -170,6 +170,183 @@ class EmptyResponseAdaptor {
+  }
+
+ private:
+  // The constructor is private because we always want to wrap the object in
+  // a shared pointer. The lifetime is controlled by any pending operations in
+  // the CompletionQueue.
+  RetryAsyncUnaryRpcFuture(char const* location,
+                           std::unique_ptr<RPCRetryPolicy> rpc_retry_policy,
+                           std::unique_ptr<RPCBackoffPolicy> rpc_backoff_policy,
+                           IdempotencyPolicy idempotent_policy,
+                           MetadataUpdatePolicy metadata_update_policy,
+                           AsyncCallType async_call, Request request)
+      : location_(location),
+        rpc_retry_policy_(std::move(rpc_retry_policy)),
+        rpc_backoff_policy_(std::move(rpc_backoff_policy)),
+        idempotent_policy_(std::move(idempotent_policy)),
+        metadata_update_policy_(std::move(metadata_update_policy)),
+        async_call_(std::move(async_call)),
+        request_(std::move(request)) {}
+
+  /// The callback for a completed request, successful or not.
+  void OnCompletion(std::shared_ptr<RetryAsyncUnaryRpcFuture> self,
+                    CompletionQueue cq, StatusOr<Response> result) {
+    if (result) {
+      final_result_.set_value(std::move(result));
+      return;
+    }
+    if (!idempotent_policy_.is_idempotent()) {
+      Status detailed_status(
+          result.status().code(),
+          FullErrorMessage("non-idempotent operation failed", result.status()));
+      final_result_.set_value(std::move(detailed_status));
+      return;
+    }
+    if (!rpc_retry_policy_->OnFailure(result.status())) {
+      std::string full_message =
+          FullErrorMessage(RPCRetryPolicy::IsPermanentFailure(result.status())
+                               ? "permanent error"
+                               : "too many transient errors",
+                           result.status());
+      Status detailed_status(result.status().code(), full_message);
+      final_result_.set_value(std::move(detailed_status));
+      return;
+    }
+    cq.MakeRelativeTimer(rpc_backoff_policy_->OnCompletion(result.status()))
+        .then([self, cq](future<std::chrono::system_clock::time_point>) {
+          self->StartIteration(self, cq);
+        });
+  }
+
+  /// The callback to start another iteration of the retry loop.
+  void StartIteration(std::shared_ptr<RetryAsyncUnaryRpcFuture> self,
+                      CompletionQueue cq) {
+    auto context =
+        ::google::cloud::internal::make_unique<grpc::ClientContext>();
+    self->rpc_retry_policy_->Setup(*context);
+    self->rpc_backoff_policy_->Setup(*context);
+    self->metadata_update_policy_.Setup(*context);
+
+    cq.MakeUnaryRpc(async_call_, request_, std::move(context))
+        .then([self, cq](future<StatusOr<Response>> fut) {
+          self->OnCompletion(self, cq, fut.get());
+        });
+  }
+
+  /// Generate an error message
+  std::string FullErrorMessage(char const* where, Status const& status) {
```

Comment:
Perhaps something that returned a `Status` (the `detailed_status`) would make it clearer that we're just prefixing the message of the input `Status`.  That is, ...
```
Status DetailedStatus(char const* where, Status const& status) {
  std::string full_message = ... + ", error=" + status.message();
  return Status(status.code(), full_message);
}
```

Reply:
Good idea, done.